### PR TITLE
CSCQVAIN-115: Local build playbook

### PR DESCRIPTION
**Pull request by tahme**
Playbook that simplifies building qvain locally.

### CODE REVIEW

File: ansible/roles/local_build/tasks/node.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+
+- name: Install nvm
+  shell: "cd; curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.34.0/install.sh | bash"
```
Comment (jppiiroinen) [inline ansible/roles/local_build/tasks/node.yml]: we should consider that we fork this repository/project, as this can be an attack vector. we download a shell script from 3rd party and execute it.

Comment (tahme) [reply ansible/roles/local_build/tasks/node.yml]: Maybe we could just directly download a specific binary tarball (with a checksum) from nodejs.org. I'm not sure if node needs any configuration or if you can just unpack and then run the binaries directly.

Comment (jppiiroinen) [reply ansible/roles/local_build/tasks/node.yml]: We could download https://github.com/nvm-sh/nvm/archive/v0.34.0.tar.gz and compare SHA hash that it is the same what it is now. so that we store the hash with the Ansible and download the full source tar ball, from which we can run the install.sh, as long as we have done a code review on the tarball.

Comment (tahme) [reply ansible/roles/local_build/tasks/node.yml]: It seems that install.sh downloads the latest version of nvm from git so there's not much security benefit from checksumming if it just gets untrusted code from the web anyway.

Comment (jppiiroinen) [reply ansible/roles/local_build/tasks/node.yml]: then we should use that nodejs as you suggested. it should work out of the box by just extracting if we have the platform binaries. what we might want to do is to add the bin folder to PATH, which I presume that nvm is doing as well

Comment (jppiiroinen) [reply ansible/roles/local_build/tasks/node.yml]: ahem, lets stop right here.

Comment (jppiiroinen) [reply ansible/roles/local_build/tasks/node.yml]: we are out of scope of this issue. the fact is that we already are having the same pattern there in another role. what we should do, instead of copying the lines, is to use that role instead. Then we can fix this issue which was spoken here with in the right scope and with different issue.

File: ansible/roles/local_build/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+
+- name: Set path variables
+  set_fact:
+    root_path: "{{ playbook_dir | dirname }}"
+    release_tree_path: "{{ playbook_dir | dirname }}/releases"
+    release_path: "{{ playbook_dir | dirname }}/releases/{{ deployment_environment_id }}"
+
+- import_tasks: go.yml       # check go version
+- import_tasks: node.yml     # install nvm to ensure our node version is recent enough
```
Comment (jppiiroinen) [inline ansible/roles/local_build/tasks/main.yml]: so here, we should change this and any other lines where we have a role for installing the stuff under ansible/roles.

File: ansible/roles/local_build/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+---
+
+- name: Set path variables
+  set_fact:
+    root_path: "{{ playbook_dir | dirname }}"
+    release_tree_path: "{{ playbook_dir | dirname }}/releases"
+    release_path: "{{ playbook_dir | dirname }}/releases/{{ deployment_environment_id }}"
+
+- import_tasks: go.yml       # check go version
```
Comment (jppiiroinen) [inline ansible/roles/local_build/tasks/main.yml]: here is another one which has a role already

File: ansible/inventories/stable/group_vars/all [hunk 4]
```diff
@@ -1,2 +1,3 @@
 deployment_environment_id: stable
-repo_branch: master
+repo_branch: next
```
Comment (jppiiroinen) [inline ansible/inventories/stable/group_vars/all]: was this change on purpose

File: ansible/local_build.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+# Clones and builds qvain-api and qvain-js locally.
+# Uses git with force so any local uncommitted
+# changes in qvain-api and qvain-js subdirs
+# are lost.
+#
+# Also creates the symlinks expected by deploy_appserver.yml
+# under qvain-ops/releases/[deployment_environment_id]
+
+- hosts: local_build
+  connection: local
```
Comment (jppiiroinen) [inline ansible/local_build.yml]: missed this on the first review round, just a comment, no need to change anything.

Comment (tahme) [reply ansible/local_build.yml]: I think I'll add `become: no` so it doesn't accidentally sudo anything.

File: ansible/inventories/stable/group_vars/all (hunk 4 above)
Comment (tahme) [reply ansible/inventories/stable/group_vars/all]: yes

**Comment by tahme** [comment]
> Add a readme.md file under ansible/roles/local_build which will explain the role and also add an example there how to use it.
readme.md added.
